hera_gpu/redcal.py

```diff
@@ -0,0 +1,398 @@
+                ubl_map[uij] = len(ubl_map)
+        ggu_indices = np.array([(gain_map[gi], gain_map[gj], ubl_map[uij]) for (gi, gj, uij) in terms], dtype=np.uint)
+        v = sol[gi]
+        shape, dtype, ndata = v.shape, v.dtype, v.size
+        assert dtype in (np.complex64, np.complex128)
+        if precision is None:
+            if dtype == np.complex64:
+                precision = 1
+            else:
+                precision = 2
+        gains = np.empty((len(gain_map), ndata), dtype=dtype)
+        for k,v in gain_map.items():
+            gains[v] = sol[k].flatten()
+        ubls = np.empty((len(ubl_map), ndata), dtype=dtype)
+        for k,v in ubl_map.items():
+            ubls[v] = sol[k].flatten()
+        data = np.array([self.data[k].flatten() for k in self.keys])
+        wgts = np.array([self.wgts[k].flatten() for k in self.keys])
+        if wgts.shape != data.shape:
+            wgts = np.resize(wgts, data.shape)
+        result = omnical(ggu_indices, gains, ubls, data, wgts, 
+            conv_crit, maxiter, check_every, check_after,
+            nthreads=NTHREADS, max_memory=MAX_MEMORY,
+            precision=precision, gain=self.gain, verbose=verbose)
+        for k,v in gain_map.items():
+            sol[k] = np.reshape(result['gains'][v], shape)
+        for k,v in ubl_map.items():
+            sol[k] = np.reshape(result['ubls'][v], shape)
+        meta = {
+            'iter': np.reshape(result['iters'], shape),
+            'chisq': np.reshape(result['chisq'], shape),
+            'conv_crit': np.reshape(result['conv'], shape),
+        }
+        return meta, sol
+
+class RedundantCalibratorGPU(RedundantCalibrator):
+    def omnical_gpu(self, data, sol0, wgts={}, gain=.3, conv_crit=1e-10, maxiter=50, check_every=4, check_after=1, precision=None):
+        """Use the Liu et al 2010 Omnical algorithm to linearize equations      and iteratively minimize chi^2.
+ 
+        Args:
+            data: visibility data in the dictionary format {(ant1,ant2,pol)     : np.array}
+            sol0: dictionary of guess gains and unique model visibilities,      keyed by antenna tuples
+                like (ant,antpol) or baseline tuples like. Gains should inc     lude firstcal gains.
+            wgts: dictionary of linear weights in the same format as data.      Defaults to equal wgts.
+            conv_crit: maximum allowed relative change in solutions to be c     onsidered converged
+            maxiter: maximum number of omnical iterations allowed before it      gives up
+            check_every: Compute convergence every Nth iteration (saves com     putation).  Default 4.
+            check_after: Start computing convergence only after N iteration     s.  Default 1.
+            gain: The fractional step made toward the new solution each ite     ration.  Default is 0.3.
+                Values in the range 0.1 to 0.5 are generally safe.  Increas     ing values trade speed
+                for stability.
```

Comment:
add `precision` to Args: in the docstring and have the order of the descriptions be the same as the order in the function.

Reply:
Done.